framework/src/materials/Material.C

```diff
@@ -171,8 +175,26 @@ Material::getOutputs()
 void
 Material::computeProperties()
 {
-  for (_qp = 0; _qp < _qrule->n_points(); ++_qp)
+  // If this Material has the _constant_on_elem flag set, we take the
+  // value computed for _qp==0 and use it at all the quadrature points
+  // in the Elem.
+  if (_constant_on_elem)
+  {
+    // Compute MaterialProperty values at the first qp.
+    _qp = 0;
     computeQpProperties();
+
+    // Copy the first value computed over to the others.
+    MaterialProperties & props = _material_data->props();
```

Comment:
I will call `retrievePropertyId()` since the FEProblem hands back a constant reference to the `Storage`.  The code is already done, IMO it's not too bad as-is.  Just adding the other test now...